Unstabilize CPU-related functionality, merge modules

bugadani opened the pull request:
Multi-core operations, as well as CPU reset is no longer stable. CPU-related operations are now contained in the `cpu` module (as well as system reset, which may be slightly awkward but keeping that alone in a separate reset module seemed equally weird).

This PR also fixed CPU reset. In esp-idf, the rom function is defined as `void esp_rom_software_reset_cpu(int cpu_no);` and we were missing the `cpu_no` parameter. Because of this change, the function is no longer marked as stable.

I think these were the last bits in esp-hal that weren't supposed to be stable.

MabezDev commented on esp-hal/src/cpu.rs:
```diff
@@ -0,0 +1,194 @@
+
+/// Performs a software reset on the chip.
+#[inline]
+pub fn software_reset() {
```
I suppose this (and the ROM function) should have a return type of  `-> !`?

MabezDev reviewed (changes requested):
Sorry to bikeshed, I feel like this would be better placed in a `system` or `soc` module.

Which also brought my attention to https://docs.esp-rs.org/esp-hal/esp-hal/0.23.1/esp32s3/esp_hal/system/index.html which probably should be hidden completely I think?

bugadani commented:
Good shout, although that system module is equally pointless IMHO Radio clocks belong to maybe `clock`, RadioPeripherals(?) and PeripheralIter belongs to `peripherals`.

MabezDev commented:
> Good shout, although that system module is equally pointless IMHO Radio clocks belong to maybe `clock`, RadioPeripherals(?) and PeripheralIter belongs to `peripherals`.

Yep, I agree with all those points!

bjoernQ commented on esp-hal/src/cpu.rs:
```diff
@@ -0,0 +1,194 @@
+    BT,
+}
+
+bitflags::bitflags! {
```
seems it's only used in `rtc_cntl` - should we move it there?

bugadani commented:
I moved everything into `system`, and removed the `EnumIter` thing that was a perma-stable-public implementation of the doc(hidden) Peripheral enum.

The Peripheral enum can and should probably be generated based on data in the soc::peripherals module, but for now it should be fine as it is.

bjoernQ reviewed (approved):
LGTM

MabezDev reviewed (approved):
LGTM, just one question that I either can't tell from the diff or has some other effect I'm not considering

bugadani commented on esp-hal/src/system.rs:
```diff
@@ -24,7 +14,7 @@ pub(crate) const KEEP_ENABLED: &[Peripheral] = &[
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
 #[repr(u8)]
 #[cfg_attr(feature = "defmt", derive(defmt::Format))]
 pub enum Peripheral {
```
It's hidden, so does it matter?

bugadani replied:
Just checked - Peripheral needs to be public, because Info in e.g. UART is public, and it needs to be public because the Instance traits refer to the Info types in a public way.

MabezDev replied:
Oops, missed the doc hidden :D, thanks!
